### PR TITLE
Tracers

### PR DESCRIPTION
**Pull request by bugaevc**
This PR:
* removes a few useful syscalls
* removes `profile(1)`
* breaks `strace(1)`

🙃

WIP because:
* I want some feedback on the general direction of this
* We need to decide what to do with `SimpleBufferBuilder`
* I would like to implement a streaming JSON parser (would would be the good API for that?), port everything to use it, unbreak `strace(1)`
* We have to expose the malloc profile events through this

More on the last item: it would still be the env variable that makes the process actually produce any data, but then the data would go to nowhere unless someone is malloc-tracing this process? How do we expose executable path and PID if, as written, tracers only support outputting an array of events?

### CODE REVIEW

File: Kernel/FileSystem/ProcFS.h [hunk 1]
```diff
@@ -108,6 +108,7 @@ class ProcFSInode final : public Inode {
     virtual KResult chmod(mode_t) override;
     virtual KResult chown(uid_t, gid_t) override;
     virtual KResultOr<NonnullRefPtr<Custody>> resolve_as_link(Custody& base, RefPtr<Custody>* out_parent = nullptr, int options = 0, int symlink_recursion_level = 0) const override;
+    virtual RefPtr<FileDescription> preopen_fd() override;
```
Comment (awesomekling) [inline Kernel/FileSystem/ProcFS.h]: This feels unrelated to the commit.

File: Kernel/FileSystem/Inode.h [hunk 2]
```diff
@@ -87,7 +87,7 @@ class Inode : public RefCounted<Inode>
     bool bind_socket(LocalSocket&);
     bool unbind_socket();
 
-    virtual FileDescription* preopen_fd() { return nullptr; };
+    virtual RefPtr<FileDescription> preopen_fd() { return nullptr; };
```
[the remark this answers is not in the document]
Comment (bugaevc) [reply Kernel/FileSystem/Inode.h]: Ah, so *that's* why. Thanks!

File: Kernel/Tracing/ProcessTracer.h [hunk 3]
```diff
@@ -47,11 +49,11 @@ class ProcessTracer : public File {
     virtual String absolute_path(const FileDescription&) const override;
 
     void did_syscall(u32 function, u32 arg1, u32 arg2, u32 arg3, u32 result);
-    pid_t pid() const { return m_pid; }
+    Process* process() const { return m_process; }
```
[the remark this answers is not in the document]
Comment (bugaevc) [reply Kernel/Tracing/ProcessTracer.h]: Not really; but it's not like we'd be const correct here anyway, we'd just `const_cast` it away at use site if we had a const tracer and wanted to mutate its process. We can go that way if you prefer that :)

[the remark this answers is not in the document]
Comment (awesomekling) [reply Kernel/Tracing/ProcessTracer.h]: For trivial getters like this, I usually add a const and a non-const version so I at least have the option of writing const-correctly :)

File: Kernel/Tracing/ProcessTracer.h [hunk 4]
```diff
@@ -61,7 +63,6 @@ class ProcessTracer : public File {
         u32 result;
     };
 
-    pid_t m_pid;
-    bool m_dead { false };
+    Process* m_process;
```
[the remark this answers is not in the document]
Comment (bugaevc) [reply Kernel/Tracing/ProcessTracer.h]: There's a problem with using a weak pointer here: weak pointer gets cleared when the pointee is destructed, but `Process` is special because of its multi-step destruction ceremony. Here, we want the process pointer cleared as soon as `Process::die()` is called, which is way before it gets finalized and finally destructed.

[the remark this answers is not in the document]
Comment (awesomekling) [reply Kernel/Tracing/ProcessTracer.h]: Ah, touché! That's a really good point. Also lol @ "multi-step destruction ceremony" which is exactly what it is.. :D

File: Kernel/FileSystem/ProcFS.h (hunk 1 above)
Comment (awesomekling) [reply Kernel/FileSystem/ProcFS.h]: Oops, I was not reading this in the right order, never mind :)

**Comment by awesomekling** [comment]
Closing this for lack of activity. Feel free to reopen if/when progress is made :)
